A revised KKS solving method

xueyang94 opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
One of the disadvantages of KKS phase-field models compared to the grand-potential models in MOOSE is that KKS solves phase concentrations (ci) as nonlinear variables, and therefore increases the system degree of freedom (DOF). If a multi-phase multi-component KKS model has highly nonlinear free energy functionals, the extra DOF causes further difficulty with the nonlinear convergence. In this implementation, ci are solved as material properties. They exist locally on the quadrature points instead of on the nodes, so the DOF is reduced and the memory usage is decreased. Using a nested Newton iteration to compute ci improves the overall convergence.

## Design
Two material classes are created: KKSPhaseConcentration(MultiPhase)Material and KKSPhaseConcentration(MultiPhase)Derivatives. The former implements a nested Newton iteration to calculate ci, and the latter solves two linear systems for the derivatives of ci with respect to the nonlinear variables, such as the global concentration and order parameters. These derivative terms are used in the KKS kernels. The KKS kernels have also been modified to accommodate ci as material properties.

## Impact
This enhancement will both change the existing API and add new classes.

refs #19980 
@dschwen  @tonkmr @jiangwen84 


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

dschwen commented:
As a general comment I would first suggest that you don't change the functionality of existing KKS kernels, but add new kernels for the local solve approach. Secondly I'd suggest you limit the scope of your first PR. Maybe add a simple two phase model first before adding a multiphase version.

The material that solves for the local concentrations should also provide updated values for the free energies as material properties, that way you don't need two copies of the free energy materials (one a compute=false, one a compute=true version).

In #20006 I'm working on a new way to cache local solve results which could potentially replace stateful phase concentrations and speed up difficult solves substantially. I'm very interested in getting a working version of the new local solve into MOOSE soon.

dschwen reviewed (commented):
Were you planning to work on this?

xueyang94 commented:
> Were you planning to work on this?

Apologies, I was catching up with another project. And yes, I'm getting back to this.

xueyang94 commented on modules/phase_field/src/materials/KKSPhaseConcentrationDerivatives.C:
```diff
@@ -0,0 +1,185 @@
+void
+KKSPhaseConcentrationDerivatives::computeQpProperties()
+{
+  // declare Jacobian matrix A
+  std::vector<std::vector<Real>> A(_num_c * 2);
+  for (auto & row : A)
+    row.resize(_num_c * 2);
+
+  // initialize all elements in A to be zero
+  for (unsigned int m = 0; m < _num_c * 2; ++m)
+  {
+    for (unsigned int n = 0; n < _num_c * 2; ++n)
+      A[m][n] = 0;
+  }
+
+  // fill in the non-zero elements in A
+  for (unsigned int m = 0; m < _num_c; ++m)
+  {
+    for (unsigned int n = 0; n < _num_c; ++n)
+    {
+      // equal chemical potential derivative equations
+      A[m * 2][n * 2] = (*_d2Fjdcjdbj[0][m][n])[_qp];
+      A[m * 2][n * 2 + 1] = -(*_d2Fjdcjdbj[1][m][n])[_qp];
+    }
+
+    // concentration conservation derivative equations
+    A[m * 2 + 1][m * 2] = 1 - _prop_h[_qp];
+    A[m * 2 + 1][m * 2 + 1] = _prop_h[_qp];
+  }
+
+  MatrixTools::inverse(A, A);
```
I can replace line 106 to 108 with NestedSolve matrix, but the rest stays the same except for changing the square brackets indexing to parenthesis indexing of matrix A, correct? I don't see how I can simplify the rest with NestedSolve in here.

dschwen replied:
You're right. My suggestion would be to simply use an Eigen::Matrix here and use Eigen to invert it. Also allocate and size all vectors and the matrix at construction time only.

xueyang94 commented on modules/phase_field/src/kernels/KKSACBulkC.C:
```diff
@@ -17,39 +17,117 @@ KKSACBulkC::validParams()
+    for (unsigned int n = 0; n < 2; ++n)
+    {
+      // declare _prop_dcideta. m is the numerator species, n is the phase of the numerator i
+      _prop_dcideta[m][n] = &getMaterialPropertyByName<Real>(_dcideta_names[m * 2 + n]);
```
`_prop_dcideta` is the output material property from KKSPhaseConcentrationDerivatives. To the KKSACBulkC kernel, `_prop_dcideta` is just a material property. We don't compute the derivative in this kernel. ci is not an explicit function of eta in the kernel.

dschwen replied:
That doesn't really matter, the point here is to automatically build the derivative property names.

xueyang94 replied:
I guess I should do the same in the KKSPhaseConcentrationDerivatives class? If the users do not supply the derivative names, how can I pass the derivative values from the material to the kernel? The `getMaterialPropertyDerivative(ci, eta)` will just be zero since the kernel does not know that ci is a function of eta.

xueyang94 commented on modules/phase_field/src/kernels/KKSACBulkC.C:
```diff
@@ -17,39 +17,117 @@ KKSACBulkC::validParams()
+      "Phase concentrations. These must have the same order as Fj_names and global_cs, for "
+      "example, c1, c2, b1, b2.");
+  params.addParam<std::vector<MaterialPropertyName>>(
+      "dcidb_names",
```
dcidb_names is not really a 'derivative' to this KKSACBulkC kernel. It is computed in KKSPhaseConcentrationDerivatives as a material property. This kernel takes dcidb as a material property as well (for use in the chain rule). The _prop_dcideta below is very similar to this.

xueyang94 commented on modules/phase_field/src/kernels/KKSACBulkF.C:
```diff
@@ -17,36 +17,97 @@ KKSACBulkF::validParams()
 Real
 KKSACBulkF::computeDFDOP(PFFunctionType type)
 {
-  const Real A1 = _prop_Fa[_qp] - _prop_Fb[_qp];
```
My bad, I'll add it back on.

dschwen commented:
We'll discuss this on Monday

xueyang94 commented on modules/phase_field/src/materials/KKSPhaseConcentrationMaterial.C:
```diff
@@ -46,17 +48,37 @@ KKSPhaseConcentrationMaterial::KKSPhaseConcentrationMaterial(const InputParamete
+      _Fi_names[0], _ci_names[0]);
+  // _dFidci[0] = &getMaterialPropertyDerivative<Real>(_Fi_names[0], _ci_names[0]);
+
+  // declare _fi_copy to be passed in to the kernels
+  for (unsigned int m = 0; m < 2; ++m)
+  {
+    _prop_fi[m] = &getMaterialPropertyByName<Real>(_Fi_names[m]);
+    _fi_copy[m] = &declareProperty<Real>(_fi_copy_name[m]);
+  }
```
@dschwen Is this how we declare the free energy copies? Line 179 assigns the free energy material to the copies.

dschwen replied:
yeah. maybe just do `_fi_copy[m] = &declareProperty<Real>(_prefix + _Fi_names[m]);` where `_prefix` is a fixed string.

xueyang94 commented on modules/phase_field/src/materials/KKSPhaseConcentrationMaterial.C:
```diff
@@ -46,17 +48,37 @@ KKSPhaseConcentrationMaterial::KKSPhaseConcentrationMaterial(const InputParamete
+  _prop_dFidci[0] = &declareProperty<Real>(_dFidci_names[0]);
+  _dFidci_names[0] = DerivativeMaterialPropertyNameInterface::derivativePropertyNameFirst(
+      _Fi_names[0], _ci_names[0]);
+  // _dFidci[0] = &getMaterialPropertyDerivative<Real>(_Fi_names[0], _ci_names[0]);
```
@dschwen Here I want to compute the free energy derivatives and generate the dFi/dci names automatically. We talked about do not directly use the DerivativeMaterialPropertyNameInterface, but use getMaterialPropertyDerivative instead. I am unsure how that works to get the derivative names...

xueyang94 commented on modules/phase_field/src/kernels/KKSACBulkF.C:
```diff
@@ -44,70 +34,68 @@ KKSACBulkF::KKSACBulkF(const InputParameters & parameters)
+    _dcideta(_num_c * 2),
+    _dcidb(_num_c * _num_c * 2)
 {
+  _prop_Fi[0] = &getMaterialPropertyByName<Real>("cp" + _Fa_name);
+  _prop_Fi[1] = &getMaterialPropertyByName<Real>("cp" + _Fb_name);
```
@dschwen In this push I added the energy Fi and derivative (dFidci and dcidc) calculations in the materials with names generated automatically, and updated the kernels. Is this how we grab the Fi from the materials? dcidc seems to be zero in the kernels even though it was declared in KKSPhaseConcentrationDerivatives. Also the pull from moose outdates my compiler so I can't compile to test right now, hopefully it is easy to fix. (I will give the kernels new names by the end.)

xueyang94 commented on modules/phase_field/src/materials/KKSPhaseConcentrationMaterial.C:
```diff
@@ -45,11 +43,12 @@ KKSPhaseConcentrationMaterial::KKSPhaseConcentrationMaterial(const InputParamete
-    _Fi_names(getParam<std::vector<MaterialPropertyName>>("Fi_names")),
-    _prop_fi(2),
-    _fi_copy(2),
+    _F1(getMaterial("Fa_material")),
+    _F2(getMaterial("Fb_material")),
```
The runtime error `To ensure dependency resolution, discrete materials must be retrieved during initial setup. This is a code problem.` is caused by these two lines. If I remove `_f1` and `_f2` entirely, that error is gone. However, I would get another runtime error `no MaterialPropertyName parameter named "Fa_material" found.`. We were using the MaterialBase and getMaterial the same way as before, strange that it is not working somehow. ("Fa_material" and "Fb_material" are defined as compute=false materials.) @dschwen

xueyang94 commented:
The MaterialBase dependency error I had was due to PR #20106 occurred on Feb 22 2022. I added an initialSetup to get the free energies objects which removed the error. I kept the original KKS kernels and added my new kernels with a "Nest" prefix. The new classes are: NestKKSSABulkF, NestKKSACBulkF, and NestKKSSplitCHCRes. The added material classes are: KKSPhaseConcentrationMaterial and KKSPhaseConcentrationDerivatives. I have also added necessary comments to all the new classes. The current code can compile and run with no issue. 

One thing I am unsure of is the `args`. My code can compute the off-diagonal Jacobian wrt coupled variables that are explicit in free energies, but not the implicit ones. I think this is the way it should be as in the original code, but I am not totally sure. @dschwen @jiangwen84 @tonkmr

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1025213/) on 3a3c167 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 3bf6ac530bb0a13e2c69331c723eea1618635519`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1025755/) on 8ebe1b5 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/19981/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1037636/) on efcd36a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 1bc5c2fc3e2548722553080dbcf48a51432fbe33`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1037676/) on 517ea54 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format e1fe79b11930943a2e4195541d0cc58bd62f0079`

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1038624/) on 2f831e0 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>6640a9</th>
      <th colspan=4>#19981 2f831e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.60%</td>
      <td>82.59%</td>
      <td>-0.00%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>75291</td>
      <td>75290</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15863</td>
      <td>15866</td>
      <td>+3</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19981/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19981/coverage/framework)


## Modules coverage

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>6640a9</th>
      <th colspan=4>#19981 2f831e</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.21%</td>
      <td>90.35%</td>
      <td>+0.14%</td>
      <td>95.08%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>12040</td>
      <td>12408</td>
      <td>+368</td>
      <td>367</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1307</td>
      <td>1325</td>
      <td>+18</td>
      <td>19</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/19981/coverage_diff/phase_field)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/19981/coverage/phase_field)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/19981/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 0.00% is less than the suggested 90.0%



This comment will be updated on new commits.

xueyang94 commented:
I will squash all the commits once review is passed.

xueyang94 commented:
@dschwen Do you prefer me to squash all the commits on this branch, or leave some "critical" ones?

hugary1995 commented:
Merging this PR also closes #21294

xueyang94 commented:
@dschwen This PR is ready to merge. Please let me know if any change is needed.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1082743/) on a9e4b00 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 74571c498fe4a3e6353b9582eff6aac0ccb28c67`

xueyang94 committed:
Two-phase Nested KKS Solve (#19981)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1110102/) on 9b5a995 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 04e909d8a3d48f651901c5b9439a9eeaebe03b75`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1110102/) on 9b5a995 wanted to post the following:

The following file(s) are changed:

  conda/mpich/conda_build_config.yaml

The conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1110102/) on 9b5a995 wanted to post the following:

The following file(s) are changed:

  conda/petsc/meta.yaml

The PETSc conda configuration has changed; ping @idaholab/moose-dev-ops

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1110102/) on 9b5a995 wanted to post the following:

The following file(s) are changed:

  conda/libmesh/meta.yaml

The libmesh conda configuration has changed; ping @idaholab/moose-dev-ops

dschwen commented on modules/phase_field/include/kernels/NestKKSACBulkF.h:
```diff
@@ -0,0 +1,81 @@
+ *
+ * The non-linear variable for this Kernel is the order parameter 'eta'.
+ */
+// class KKSACBulkF : public Kernel
```
```suggestion
```

xueyang94 replied:
I would be happy to implement the KKS action at some point.

dschwen commented on modules/phase_field/src/materials/KKSPhaseConcentrationDerivatives.C:
```diff
@@ -0,0 +1,175 @@
+{
+  // declare Jacobian matrix A
+  Eigen::MatrixXd A(_num_c * 2, _num_c * 2);
+
+  // initialize all elements in A to be zero
+  for (unsigned int m = 0; m < _num_c * 2; ++m)
+  {
+    for (unsigned int n = 0; n < _num_c * 2; ++n)
+      A(m, n) = 0;
+  }
```
```suggestion
  A.setZero()
```

should work here

dschwen reviewed (dismissed):
This looks nice!

At some point we should add (or flesh out ) the KKS action to set up these systems easily.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1163778/) on c7f97ee wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fb387056f8b889468cab177a1a7dffb3671f7958`

dschwen commented:
You just added a merge commit (c7f97ee51e61cf62070d5642d7cec0a74dc1114d) that you need to get rid of.

xueyang94 commented:
> You just added a merge commit ([c7f97ee](https://github.com/idaholab/moose/commit/c7f97ee51e61cf62070d5642d7cec0a74dc1114d)) that you need to get rid of.

I just applied your suggested changes. I will squash all commits once this check is passed.

xueyang94 commented:
> You just added a merge commit ([c7f97ee](https://github.com/idaholab/moose/commit/c7f97ee51e61cf62070d5642d7cec0a74dc1114d)) that you need to get rid of.

Why does the merge cause the App documentation check to fail? I can't see the detailed error message...

dschwen commented:
> > You just added a merge commit ([c7f97ee](https://github.com/idaholab/moose/commit/c7f97ee51e61cf62070d5642d7cec0a74dc1114d)) that you need to get rid of.
> 
> Why does the merge cause the App documentation check to fail? I can't see the detailed error message...

The merge commit doesn't have anything to do with the app failure. I invalidated that test, but you should still remove teh merge commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1164991/) on 6da1ab6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fa664410cea12f16d74333e1a608071309ee9ef2`

xueyang94 commented:
> > > You just added a merge commit ([c7f97ee](https://github.com/idaholab/moose/commit/c7f97ee51e61cf62070d5642d7cec0a74dc1114d)) that you need to get rid of.
> > 
> > 
> > Why does the merge cause the App documentation check to fail? I can't see the detailed error message...
> 
> The merge commit doesn't have anything to do with the app failure. I invalidated that test, but you should still remove teh merge commits.

The merge commit has been removed.

xueyang94 committed:
Changed the kks test to be the local solve version of the existing kks split example (#19981)

xueyang94 commented:
@dschwen I have changed the test to be the local solve version of the existing KKS example located at modules/phase_field/test/tests/KKS_system/kks_example_split.i. Is there any other change needed before merging?

dschwen commented on modules/phase_field/include/kernels/NestKKSACBulkC.h:
```diff
@@ -0,0 +1,66 @@
+
+#include "KKSACBulkBase.h"
+
+// Forward Declarations
+
```
```suggestion
```

dschwen commented on modules/phase_field/include/kernels/NestKKSACBulkC.h:
```diff
@@ -0,0 +1,66 @@
+ *
+ * The non-linear variable for this Kernel is the order parameter 'eta'.
+ */
+
```
```suggestion
```

dschwen commented on modules/phase_field/doc/content/source/kernels/NestKKSACBulkC.md:
```diff
@@ -0,0 +1,65 @@
+# NestKKSACBulkC
```
Would you mind renaming everything `Nest...` to `Nested..` ?

dschwen commented on modules/phase_field/include/kernels/NestKKSACBulkF.h:
```diff
@@ -0,0 +1,81 @@
+
+#include "KKSACBulkBase.h"
+
+// Forward Declarations
+
```
```suggestion
```

dschwen commented on modules/phase_field/include/kernels/NestKKSACBulkF.h:
```diff
@@ -0,0 +1,81 @@
+ *
+ * The non-linear variable for this Kernel is the order parameter 'eta'.
+ */
+
```
```suggestion
```

dschwen commented on modules/phase_field/include/kernels/NestKKSSplitCHCRes.h:
```diff
@@ -0,0 +1,77 @@
+ * The user picks one phase free energy \f$ F_1 \f$ (f_base) and its
+ * corresponding phase concentration \f$ c_1 \f$
+ */
+
```
```suggestion
```

dschwen commented on modules/phase_field/include/kernels/NestKKSSplitCHCRes.h:
```diff
@@ -0,0 +1,77 @@
+  const std::vector<MaterialPropertyName> _c1_names;
+
+  /// Free energy
+  const MaterialPropertyName _F1_name;
```
You are using a mix of F1 and Fa in these classes (this might come from the old KKS code). I think it would be clearer if we could pick one notation.

dschwen commented on modules/phase_field/src/kernels/NestKKSACBulkC.C:
```diff
@@ -0,0 +1,151 @@
+  params.addClassDescription("KKS model kernel (part 2 of 2) for the Bulk Allen-Cahn. This "
+                             "includes all terms dependent on chemical potential.");
+  params.addRequiredCoupledVar("global_cs", "The interpolated concentrations c, b, etc");
+  params.addRequiredParam<std::vector<MaterialPropertyName>>(
+      "ci_names",
+      "Phase concentrations. The order must match Fa, Fb, and global_c, for example, c1, "
+      "c2, b1, b2, etc");
```
we can leave this as is for now, but we do have a `std::vector<std::vector<T>>`  parameter type (user input would look like `'c1 c2; b1 b2'`

xueyang94 replied:
With `std::vector<std::vector<T>>` and user input of ci_names `c1 c2; b1 b2`, do we refer to `b1` as `_ci_names[1][0]`? I would probably leave it as is so we don't need to change all the for() loops.

dschwen reviewed (commented):
Looks really good. Some minor comments to be addressed but then it's ready for merging.

xueyang94 committed:
Rename classes Nest to Nested (#19981)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1267752/) on d5d37dd wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/19981/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 13d77bc69b9b63e51e8e814ec3ba5afed206d5bd`

xueyang94 committed:
Changed Fa_name and Fa_material to fa_name (#19981)

dschwen commented on modules/phase_field/test/tests/KKS_system/kks_nested_Newton_nonconvergence.i:
```diff
@@ -0,0 +1,230 @@
+[]
+
+[Outputs]
+  file_base = kks_nested_Newton_nonconvergence
```
This input and output is not used. I'll pull the branch down to see if it makes sense including them for now to be added in a future test.

dschwen commented:
I removed the unused (and out of date) files, resolved a merge conflict, and squashed the commits.

dschwen reviewed (approved):
Looks good to me now

xueyang94 commented:
@dschwen @tonkmr Thanks for the review and merge. I have another pull request #21359 that handles multiphase models. Would you be able to check and merge that as well if needed?
